### PR TITLE
fungible/s decrease balance preserves account

### PR DESCRIPTION
**Pull request by muharem**
Ensure `preservation` is respected when decreasing balance through `fangible/s::Unbalanced::decrease_balance`.

At present, an account might be removed even if `preservation` is set to `Preserve` or `Protect`. Given the default implementation of `decrease_balance`, we can't pinpoint the exact constraint that wasn't satisfied (frozen or ED). As a result, we return the `FundsUnavailable` error.

There's a more extensive PR that aims to address the same problem: https://github.com/paritytech/polkadot-sdk/pull/1296. Due to its broader scope, I'm maintaining this one separately, in the event it can be merged sooner. The new tests introduced here will eventually be superseded by the conformance tests from the referenced PR.

Closes: https://github.com/paritytech/polkadot-sdk/issues/1698

### CODE REVIEW

**Comment by liamaharon** [comment]
I have a fix and tests for what appear to be the same thing you're working on here: https://github.com/paritytech/polkadot-sdk/pull/1296/files#diff-e04f7799b91b4225b1923b49d8484495f7868868ab32ed78919f310ffe268fd9R190-R193

**Comment by muharem** [comment]
> I have a fix and tests for what appear to be the same thing you're working on here: https://github.com/paritytech/polkadot-sdk/pull/1296/files#diff-e04f7799b91b4225b1923b49d8484495f7868868ab32ed78919f310ffe268fd9R190-R193

sure, lets work on your PR

File: substrate/frame/assets/src/tests.rs [hunk 1]
```diff
@@ -1775,3 +1775,77 @@ fn asset_destroy_refund_existence_deposit() {
...
+
+#[test]
+fn transfer_and_preserve_account() {
+	new_test_ext().execute_with(|| {
+		use frame_support::traits::tokens::Preservation::Preserve;
+
+		let id = 42;
+		Balances::make_free_balance_be(&1, 10);
+		Balances::make_free_balance_be(&2, 10);
+		assert_ok!(Assets::create(RuntimeOrigin::signed(1), id, 1, 10));
```
Comment (liamaharon) [inline substrate/frame/assets/src/tests.rs:1787]: Is calling `create` necessary? I generally just call `mint` directly on uninitialised accounts.

**Review by liamaharon — APPROVED** [review]

**Review by jsidorenko — APPROVED** [review]

File: substrate/frame/support/src/traits/tokens/fungible/regular.rs [hunk 2]
```diff
@@ -181,8 +181,9 @@ pub trait Unbalanced<AccountId>: Inspect<AccountId> {
 	) -> Result<Self::Balance, DispatchError> {
 		let old_balance = Self::balance(who);
 		let free = Self::reducible_balance(who, preservation, force);
-		if let BestEffort = precision {
-			amount = amount.min(free);
+		match precision {
+			BestEffort => amount = amount.min(free),
+			Exact => ensure!(free >= amount, TokenError::FundsUnavailable),
 		}
 		let new_balance = old_balance.checked_sub(&amount).ok_or(TokenError::FundsUnavailable)?;
```
Comment (juangirini) [inline substrate/frame/support/src/traits/tokens/fungible/regular.rs:188]: We shouldn't need this check anymore, as `amount` <= `free` <= `old_balance` (by definition `reducible_balance` must be always <= than `balance`)
https://github.com/paritytech/polkadot-sdk/blob/676bacd7299826fb0ea88055d2108f3c6a734291/substrate/frame/support/src/traits/tokens/fungible/regu
lar.rs#L80-L90
Same applies to `fungibles`

**Comment by muharem** [comment]
closing this PR, since https://github.com/paritytech/polkadot-sdk/pull/1296 is merged
thanks to @liamaharon

**Pull request closed without merging**